neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,260 @@
+---
+NEP: 519
+Title: Yield Execution
+Authors: Akhi Singhania <akhi3030@gmail.com>; Saketh Are <saketh@near.org>
+Status: Draft
+DiscussionsTo: https://github.com/near/NEPs/pull/519
+Type: Protocol
+Version: 0.0.0
+Created: 2023-11-17
+LastUpdated: 2023-11-20
+---
+
+## Summary
+
+Today, when a smart contract is called by a user or another contract, it has no sensible way to delay responding to the caller till it has observed another future transaction.  This proposal introduces this possibility into the NEAR protocol.
+
+## Motivation
+
+There exist some situations where when a smart contract on NEAR is called, it will only be able to provide an answer at some arbitrary time in the future.  So the callee needs a way to defer replying to the caller till this time in future.
+
+Examples include when a smart contract (`S`) provides MPC signing capabilities parties external to the NEAR protocol are computing the signature.  The rough steps are:
+
+1. Signer contract provides a function `fn sign_payload(Payload, ...)`.
+2. When called, the contract defers replying to the caller.
+3. External indexers are monitoring the transactions on the contract; they observe the new signing request, compute a signature, and call another function `fn signature_available(Signature, ...)` on the signer contract.
+4. The signer contract validates the signature and if validate, replies to the original caller.
+
+Today, the NEAR protocol has no sensible way to defer replying to the caller in step 2 above.  This proposal proposes adding two following new host functions to the NEAR protocol:
+
+- `yield`: this can be called by a contract to indicate to the protocol that it is not ready yet to reply to its caller.
+- `resume`: a contract can use this mechanism to indicate to a protocol that it is now ready to reply to a caller that it had deferred earlier.
+
+If these two host functions were available, then `yield` would be used in step 2 above and `resume` would be used in step 4 above.
+
+## Specification
+
+The proposal is to add the following host functions to the NEAR protocol:
+
+
+```rust
+/// Instructs the protocol that the smart contract is not yet ready to respond
+/// to its caller yet.  The smart contract promises to call
+/// `promise_yield_resume()` within X (TBD) blocks.  When
+/// `promise_yield_resume()` is called, the protocol will call the method on the
+/// smart contract that is identified by `method_name_len` and `method_name_ptr`
+/// and this method may or may not respond to the caller.
+///
+/// `arguments_len` and `arguments_ptr` provide an initial blob of arguments
+/// that will be passed to the method.
+///
+/// If the contract fails to call `promise_yield_resume()` within X blocks, then
+/// the protocol will call the method with a timeout error.
```

Comment:
In the reference implementation the method is called without any promise results (i.e. the `promise_result_count` host function will return 0). It will be up to the contract to interpret this missing data as a timeout error (probably there will be convenience macros/functions in `near-sdk` for this). Again I think it is important to be clear in the documentation about how developers will interact with this feature.

Reply:
In the reference implementation in the case of a timeout `promise_result_count()` will still be 1  and `promise_result(0)` will return [PromiseResult::Failed](https://github.com/near/near-sdk-rs/blob/9bb16c2d08d522baf52ddd126e3fe0f01386f2ca/near-sdk/src/types/vm_types.rs#L20). I would consider adjusting the documentation to reflect that there is no specific "timeout error," just a generic promise failure.

Reply:
Thanks for the correction @saketh-are !

Reply:
@saketh-are: is your recommendation then to change the sentence to something like the following:

"If the contract fails to call `promise_yield_resume()` within X blocks, then the protocol will call the method with a promise failure."

Reply:
Really we guarantee the following:
- If `promise_yield_resume` is called before the yield expires, the callback will be executed with `PromiseResult::Successful` containing the payload passed to yield_resume. 
- If resume is not called by the time the yield expires, the callback will be executed with a `PromiseResult::Failed`.
- The yield will not expire within X blocks, but it is at the discretion of the protocol exactly when it will expire after X blocks have passed.